### PR TITLE
Allow updating with just a USB connection

### PR DESCRIPTION
**Pull request by johnnyman727**
Dependent on: https://github.com/tessel/openwrt-tessel/pull/47

### CODE REVIEW

**Comment by johnnyman727** [comment]
On weird issue I'm facing is that running `sysupgrade /tmp/openwrt.bin` does not seem to work. The only thing it seems to do is kill the SPI Daemon permanently so perhaps it is beginning to restart? @kevinmehall what do you think could be happening to keep `sysupgrade` from completing when I `exec` it from the USB Daemon?

**Comment by kevinmehall** [comment]
Try adding `spid` and `usbexecd` to the whitelist of processes that sysupgrade won't kill when starting the update: https://github.com/tessel/openwrt/blob/2015-10-26/package/base-files/files/lib/upgrade/common.sh#L128.

**Comment by kevinmehall** [comment]
You can just modify `/lib/upgrade/common.sh` on a Tessel, and if it works then patch the linked file in the openwrt repo.

**Comment by johnnyman727** [comment]
Thanks, I'll give that a try.

**Comment by johnnyman727** [comment]
Unfortunately, modifying `/lib/upgrade/common.sh` didn't seem to work. 

Strangely, `ubus call system upgrade` is a `ubus` option but it doesn't seem to do anything either.

But I *could* run either `require('child_process').exec('sysupgrade /tmp/openwrt.bin')` and `require('child_process').spawn('sysupgrade, ['/tmp/openwrt.bin'])` with positive results so I guess we could do that worst case...

File: bin/tessel-2.js [hunk 1]
```diff
@@ -289,6 +289,17 @@ makeCommand('update')
     flag: true,
     help: 'Update to the latest version regardless of current version.'
   })
+  .option('openwrtPath', {
+    abbr: 'op',
+    required: false,
+    flag: false,
+    help: 'Update with the OpenWRT image at the indicated local path.'
+  })
+  .option('firmwarePath', {
```
Comment (rwaldron) [inline bin/tessel-2.js]: We should avoid camel casing command line options :P

File: lib/controller.js [hunk 2]
```diff
@@ -789,117 +789,126 @@ controller.printAvailableUpdates = function() {
...
+    }
+  });
+};
 
-      return updates.requestBuildList().then(function(builds) {
+controller.updateWithLocalBuilds = function(opts, tessel) {
+  return updates.loadLocalBinaries(opts)
+    .then((images) => {
+      return tessel.update(images);
+    })
```
Comment (rwaldron) [inline lib/controller.js]: Can just be `(images) => tessel.update(images)`

File: lib/tessel/update.js [hunk 3]
```diff
@@ -53,33 +64,36 @@ Tessel.prototype.updateOpenWRT = function(image) {
...
         });
       });
     });
 };
 
 Tessel.prototype.updateFirmware = function(image) {
-  logs.info('Updating firwmare (2/2)');
-  // This must be USB connection
+  return new Promise((resolve, reject) => {
+    logs.info('Updating firmware (2/2)');
```
Comment (rwaldron) [inline lib/tessel/update.js]: Why does this happen in the promise?

Comment (johnnyman727) [inline lib/tessel/update.js]: I was just having issues with Promises resolving when I thought they should. I can revert this change if need be.

**Comment by johnnyman727** [comment]
I think something else must be going on because I was able to successfully sysupgrade from the CLI with [this patch](https://github.com/tessel/t2-cli/commit/be9338d942467e84b315e43316b9602521658d65) and the `t2-cli usb-exec sysupgrade /tmp/openwrt.bin`... That should be essentially exactly the same as what happens in the `t2 update` command so there is something else causing the issue.

**Comment by johnnyman727** [comment]
I've confirmed that two changes are necessary:
1. The patch as outlined by @kevinmehall above
2. Not closing the child process that `exec`ed the `sysupgrade` for about two minutes (or until it reports that the upgrade completed). 

Because I did not implement 2, it seemed as though 1 made no difference but with both implementations, the update succeeds.

**Comment by johnnyman727** [comment]
Working on the OpenWRT patch [here](https://github.com/tessel/openwrt/commit/1c1896abbed88b7d10e0a3dc5466c288bb05f093) but the documentation on patches is pretty lacking. @kevinmehall can you look it over for me when you get a chance?

**Comment by kevinmehall** [comment]
The file is in the openwrt tree, so just modify it in a normal commit. Patches are used to make changes to 3rd-party code that is downloaded as an upstream tarball and doesn't live in the repository. (The directory where you added that patch is supposed to be for patches against the uClibc source)

**Comment by johnnyman727** [comment]
Ah okay, that makes much more sense. Thanks for the clarification.

**Comment by johnnyman727** [comment]
So this feature introduces a fun complication which is that because we have to update our OpenWRT build to make updating over USB work, all previous builds will not have this capability. That means we have to check the version running on the Tessel and confirm that it's > 0.0.11 before moving forward with a *USB-only* update.

@rwaldron any thoughts on that? It's kind of shitty but I'm not sure we have a good way to get around it.

**Comment by kevinmehall** [comment]
If you want a hack, the CLI could modify `/lib/upgrade/common.sh` on old versions before performing the update...

**Comment by johnnyman727** [comment]
@kevinmehall I'm ashamed at you for even suggesting such a dirty hack.

But yeah, does `sed '/*procd*\|*ash*\|*init*/c \ \*procd*|*ash*|*init*|*watchdog*|*ssh*|*dropbear*|*telnet*|*login*|*hostapd*|*wpa_supplicant*|*nas*|*spid*|*usbexecd*) : ;;'\
 /lib/upgrade/common.sh `

**Comment by kevinmehall** [comment]
<a href=http://xkcd.com/1197></a>

Don't you need `sed -i` for in-place edit? Or write to a temp file and `mv` over the original if `busybox` sed doesn't have that option.

If you were to include the whole original line, it would be a no-op in the case that the update had already been applied. Also, watch the escaping of the `*` characters in addition to the `|`.

**Comment by johnnyman727** [comment]
> Don't you need sed -i for in-place edit? Or write to a temp file and mv over the original if busybox sed doesn't have that option.

Yes, I do and busybox sed does support it. 

> If you were to include the whole original line, it would be a no-op in the case that the update had already been applied. Also, watch the escaping of the * characters in addition to the |.

Ugh I don't think my fingers can type that many escapes. It's not physically possible.

**Comment by johnnyman727** [comment]
For anyone who's checking in on this PR, it's currently blocked by my effort to add a hacky to fix an issue [described here](https://github.com/tessel/t2-cli/pull/595#issuecomment-193591352) so that folks running older versions of the T2 firmware can do updates over USB.

[The command to overwrite a specific line](https://github.com/tessel/t2-cli/blob/jon-update-usb/lib/tessel/commands.js#L160-L165) of a file which determines how the update process works doesn't seem to be effective over the USB Daemon but it does work on the Tessel shell console. I'm guessing the escapes aren't being handled properly...

**Comment by rwaldron** [comment]
There must be an easier way to do that :\

**Comment by johnnyman727** [comment]
I've tried all kinds of escaping with no luck. I may go the kinda hackier route and just ship the file with the CLI, then write it to a temporary file on T2, then rename the temporary file to the real file name (thus replacing it).

File: lib/tessel/update.js [hunk 4]
```diff
@@ -27,59 +28,108 @@ Tessel.prototype.fetchCurrentBuildInfo = function() {
...
+          if (err) {
+            return reject(err);
+          }
 
-        logs.info('Transferring image of size', (image.length / 1e6).toFixed(2), 'MB');
-        // When we finish writing the image
-        remoteProc.once('close', resolve);
-        // Write the image
-        remoteProc.stdin.end(image);
+          logs.info('Transferring image of size', (image.length / 1e6).toFixed(2), 'MB. This will take 2-4 minutes...');
```
Comment (rwaldron) [inline lib/tessel/update.js]: Use template string

**Comment by rwaldron** [comment]
Damn dude, that was smoooooooth


```
$ t2 update
INFO Looking for your Tessel...
INFO Connected to Tessel-02A347EF35F2.
INFO New firmware version found...0.0.12
INFO Updating Tessel-02A347EF35F2 to latest version (0.0.12)...
INFO Beginning update download. This could take a couple minutes..
  downloading [====================] 100% 0.0s remaining
INFO Download complete!
INFO Updating OpenWRT (1/2)
INFO Transferring image of size 19.14 MB. This will take 2-4 minutes...
INFO Transfer complete.
INFO Starting OpenWRT update.
INFO Please do not remove power from Tessel.
INFO This process will take at least two minutes...
INFO Updating firmware (2/2)
INFO Firmware update complete!
INFO Updated Tessel-02A347EF35F2 from  0.0.11  to  0.0.12
```


First time, with new Tessel 2 out of the box.

**Comment by rwaldron** [comment]
Unplugged, re-plugged, ran t2 wifi setup and we're in business.

Please address the nit-picks and I will land this tomorrow

File: bin/tessel-2.js (hunk 1 above)
Comment (johnnyman727) [inline bin/tessel-2.js]: :+1:

File: lib/controller.js (hunk 2 above)
Comment (johnnyman727) [inline lib/controller.js]: beautiful

**Comment by johnnyman727** [comment]
Nit-picks addressed!

**Pull request merged**